### ISSUE TITLE
Test case failing after PR #1212 (huge allocations for a small array)

### ISSUE THREAD

**Issue by pawosm-arm** [flang-compiler/flang#1253]
The mw4 test case which operates on just a 2-element array makes way too large allocations after merging PR #1212. Compare these (mw4-safe was compiled before merging this PR):
```
$ memusage -m ./mw4-safe
 ok

Memory usage summary: heap total: 104294, heap peak: 104230, stack peak: 1808
         total calls   total memory   failed calls
 malloc|         10         104294              0
realloc|          0              0              0  (nomove:0, dec:0, free:0)
 calloc|          0              0              0
   free|          2             64
mmap(r)|          0              0              0
mmap(w)|          0              0              0
mmap(a)|          0              0              0
 mremap|          0              0              0  (nomove: 0, dec:0)
 munmap|          0              0              0
Histogram for block sizes:
    0-15              5  50% ==================================================
   32-47              2  20% ====================
12288-12303           1  10% ==========
19200-19215           1  10% ==========
   large              1  10% ==========
$ memusage -m ./mw4
 ok

Memory usage summary: heap total: 9257652166, heap peak: 9257632928, stack peak: 2112
         total calls   total memory   failed calls
 malloc|         11     9257652166              0
realloc|          0              0              0  (nomove:0, dec:0, free:0)
 calloc|          0              0              0
   free|          2     9257547904
mmap(r)|          0              0              0
mmap(w)|          0              0              0
mmap(a)|          0              0              0
 mremap|          0              0              0  (nomove: 0, dec:0)
 munmap|          0              0              0
Histogram for block sizes:
    0-15              5  45% ==================================================
   32-47              2  18% ====================
12288-12303           1   9% ==========
19200-19215           1   9% ==========
   large              2  18% ====================
```
This causes failure in our CI where memory restrictions do apply.

**Comment by pawosm-arm** [comment]
As it turned out, the problem with the code added by upstream PR#1212 is with the call to the `move_range_after()` function in the `semutil2.c` file.

The `move_range_before()` and `move_range_after()` functions were introduced by Gary Klimowicz a couple of years ago and never actually used, even in the commit that has introduced them. Therefore there is no good example on how this function can be used legally.

Apparently, the way it is used in the code added by PR#1212 is not entirely legal. Moving some of the AST nodes ahead in the STD list without checking for possible dependencies turned out to be problematic. In the mw4 test case it resulted in a huge allocation as the computation of the allocation size was moved after the actual allocation, therefore some random value was used and when it was huge, it resulted in the allocation failure. Note that there is potential for any other failure caused by such careless motion of the AST nodes. Unfortunately, not making this move also results in an incorrect execution, so this call to `move_range_after()` can not be simply removed.

As this is issue is potentially harmful in some very non-obvious ways, this needs to be addressed.

I have worked out a temporary hack (which is not a final solution). After the move is done, the affected nodes are copied back to their original place, but only in the case when there is anything to copy (begin != end, note that in opposite case the move_range_before() function does not actually move anything) and only when the 'implied do' element count is known at compile time (otherwise, the `implied_do7` test case is failing).

**Comment by bryanpkc** [comment]
@pawosm-arm Someone from our side has been working on a fix. He just hasn't updated this issue yet.

**Comment by zhaochuanfeng** [comment]
Sorry, the other day, I thought my patch could fix this regression, but it wasn't completely fixed, I will work on it and fix it ASAP.

**Comment by pawosm-arm** [comment]
Fixed by #1261
